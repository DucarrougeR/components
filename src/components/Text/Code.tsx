import styled from '../../style/styled_components'
import { Text } from '../Text/Text'

export const Code = styled(Text)`
  font-family: ${p => p.theme.fontFaces.codeFont};
`.withComponent('code')
