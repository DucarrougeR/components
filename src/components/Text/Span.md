```js noeditor
const StatusAndResources = require('../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/SmIM3VUQUgo8xb4o7UUxflkT/Typography"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Text/Span.tsx"
  feedbackTitle="Span Component Feedback" />
```
