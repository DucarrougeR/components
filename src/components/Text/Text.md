```js noeditor
const StatusAndResources = require('../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/SmIM3VUQUgo8xb4o7UUxflkT/Typography"
  githubURL="https://github.com/looker/relens/blob/master/src/components/Text/Text.tsx"
  feedbackTitle="Text Component Feedback" />
```


### Text Element

By default the `<Text />` component will render text wrapped in a `<div>`,  if you need another html element for semantic purposes you can use the `element` property

```js
<Text>I get wrapped in a div</Text>
<Span>Wrapped in a span</Span>
<Paragraph>Wrapped in a paragraph tag</Paragraph>
<Code>Wrapped in a code tag</Code>
```

<div class="doc-section-divider"></div>

### Text Size

If you need to adjust the font-size of the rendered text, you can use the `size` property and a value from the [type scale](/#!/Typography)

```js
 <Text size="d1">Sized to d1</Text>
 <Text size="one">Sized to 1</Text>
 <Text size="six">Sized to 6</Text>
```

<div class="doc-section-divider"></div>

### Weight & Transform

Common patterns for text is to adjust the font weight and transform the text. Below is an example of using the `weight` and `textTransform` properties to modify the rendered text

```js
  <Text size="six" textTransform="upper" weight="semiBold">A great story</Text>
  <Text size="d3" weight="light">This is a great headline</Text>
  <Text size="six" textTransform="caps">Some metadata about this story</Text>
```

<div class="doc-section-divider"></div>

### Text Alignment

You can use the `align` property to change the alignment of the rendered text, below is an example with aligned text put into `<Card />`s.

```js
<div>
  <Card>
    <CardContent>
      <Text> I am aligned left by default</Text>
    </CardContent>
  </Card>
  <Card>
    <CardContent>
      <Text align="center">This is how you can center align text</Text>
    </CardContent>
  </Card>
  <Card>
    <CardContent>
      <Text align="right">This is how you can right align text</Text>
    </CardContent>
  </Card>
</div>
```

<div class="doc-section-divider"></div>

### Text Mode

Lens provides a few modes for the `<Text />` component based on common patterns we use in our applications, the `mode` property will change that will change the color the rendered text.

```js
<Text>Regular Text</Text>
<Text variant="secondary">Secondary Text</Text>
<Text variant="subdued">Subdued Text</Text>
<Text variant="critical">Critical Text</Text>
<Text variant="positive">Positive Text</Text>
```


<div class="doc-section-divider"></div>

### Truncate

At times you may want your `<Text />` to truncate instead of the text wrapping, the `truncate` property will do that for you.

```js
<div style={{width: "40%"}}>
<Text truncate>Hello there I am some text that will truncate üçïü•ëüç™ü•ì</Text>
</div>
```
