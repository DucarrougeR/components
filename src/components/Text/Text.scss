@import '../../styles/variables';

.lens-text {
  text-rendering: optimizeLegibility;
  font-family: $brand-font;
  font-weight: $default-font-weight;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-smoothing: antialiased;
}

p {
  margin: 0;
}

.is-display-1 {
  font-size: font-size(d1);
  line-height: line-height(d1);
}

.is-display-2 {
  font-size: font-size(d2);
  line-height: line-height(d2);
}

.is-display-3 {
  font-size: font-size(d3);
  line-height: line-height(d3);
}

.is-heading-1 {
  font-size: font-size(1);
  line-height: line-height(1);
}

.is-heading-2 {
  font-size: font-size(2);
  line-height: line-height(2);
}

.is-heading-3 {
  font-size: font-size(3);
  line-height: line-height(3);
}

.is-heading-4 {
  font-size: font-size(4);
  line-height: line-height(4);
}

.is-heading-5 {
  font-size: font-size(5);
  line-height: line-height(5);
}

.is-heading-6 {
  font-size: font-size(6);
  line-height: line-height(6);
}

.font-weight-light {
  font-weight: 300;
}

.font-weight-normal {
  font-weight: 400;
}

.font-weight-semi-bold {
  font-weight: 600;
}

.font-weight-bold {
  font-weight: 700;
}

.font-weight-extra-bold {
  font-weight: 600;
}

.transform-upper {
  text-transform: uppercase;
}

.transform-lower {
  text-transform: lowercase;
}

.transform-caps {
  text-transform: capitalize;
}

.transform-none {
  text-transform: none;
}

.is-secondary {
  color: color('gray', 600);
}
.is-subdued {
  color: color('gray', 500);
}

.is-positive {
  color: green;
}

.is-critical {
  color: red;
}

.is-truncated {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.is-code {
  font-family: $code-font;
}
