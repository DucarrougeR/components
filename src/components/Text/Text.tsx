import { TextDecorationProperty } from 'csstype'
import * as React from 'react'
import { css, styled, withTheme } from '../../style'
import { ThemedProps } from '../../types'
import { Box, BoxPropsWithout } from '../Box'

export type TextTransforms = 'caps' | 'lower' | 'none' | 'upper'
export type TextAlignments = 'left' | 'center' | 'right'
export type TextElements = 'span' | 'p' | 'code'
export type TextVariants =
  | 'critical'
  | 'positive'
  | 'secondary'
  | 'subdued'
  | 'inverted'

export interface TextProps
  extends BoxPropsWithout<HTMLSpanElement, 'wrap' | 'is'> {
  /** Base html text element
   *  @default "span"
   */
  is?: TextElements
  /** Align text */
  align?: TextAlignments
  /** Set text decoration property */
  decoration?: TextDecorationProperty
  /** Size mapping from type ramp */
  textTransform?: TextTransforms
  /** Adjust style of text with more meaning by using an intent */
  variant?: TextVariants
  /** Should browser insert line breaks within words to prevent text from overflowing its content box  */
  wrap?: boolean
  /** Custom css class */
  className?: string
}

function getTextTransform(transform: TextTransforms | undefined) {
  switch (transform) {
    case 'upper':
      return css`
        text-transform: uppercase;
      `
    case 'lower':
      return css`
        text-transform: lowercase;
      `
    case 'caps':
      return css`
        text-transform: capitalize;
      `
    case 'none':
    default:
      return css`
        text-transform: none;
      `
  }
}

function textVariant(props: ThemedProps<TextProps>) {
  switch (props.variant) {
    case 'critical':
      return css`
        color: ${props.theme.colors.palette.red500};
      `
    case 'positive':
      return css`
        color: ${props.theme.colors.palette.green500};
      `
    case 'secondary':
      return css`
        color: ${props.theme.colors.palette.charcoal500};
      `
    case 'subdued':
      return css`
        color: ${props.theme.colors.palette.charcoal400};
      `
    case 'inverted':
      return css`
        color: ${props.theme.colors.palette.textInverted};
      `
    default:
      return false
  }
}

function getWrap(doWrap: boolean) {
  if (doWrap) {
    return css`
      overflow-wrap: break-word;
    `
  }
  return ``
}

const InternalText: React.FC<ThemedProps<TextProps>> = ({
  is = 'span',
  align,
  decoration,
  textTransform,
  variant,
  wrap,
  theme,
  lineHeight,
  fontSize,
  fontWeight,
  ...props
}) => {
  return (
    <Box
      is={is}
      fontSize={fontSize}
      fontWeight={fontWeight}
      lineHeight={lineHeight || fontSize}
      textAlign={align}
      {...props}
    >
      {props.children}
    </Box>
  )
}

export const Text = styled<TextProps>(withTheme(InternalText))`
  text-decoration: ${props => props.decoration};
  ${props => getTextTransform(props.textTransform)};
  ${props => getWrap(props.wrap || false)};
  ${textVariant};
`
