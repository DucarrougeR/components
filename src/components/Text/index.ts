export * from './Code'
export * from './Paragraph'
export * from './Span'
export * from './Text'
