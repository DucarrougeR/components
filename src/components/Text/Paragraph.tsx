import * as React from 'react'
import { shouldTruncate, styled } from '../../style'
import { Text, TextProps } from '../Text/Text'

export interface ParagraphProps extends TextProps {
  /** Truncate text on overflow */
  truncate?: boolean
}

const InternalParagraph: React.SFC<ParagraphProps> = ({
  truncate,
  ...props
}) => {
  return (
    <Text is="p" {...props}>
      {props.children}
    </Text>
  )
}
export const Paragraph = styled<ParagraphProps>(InternalParagraph)`
  ${props => shouldTruncate(props.truncate || false)};
`
