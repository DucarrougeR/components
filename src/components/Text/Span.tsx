import * as React from 'react'
import styled, { css, StyledComponentClass } from '../../styled_components'
export { StyledComponentClass }
import { ThemeInterface } from '../../themes'
export { ThemeInterface }
// End Typescript component boilerplate

import { Text, TextProps } from '../Text/Text'

export const Span = Text.withComponent('span')
