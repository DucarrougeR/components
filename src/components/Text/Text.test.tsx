import { Text } from './Text'

import {
  snapshotTestTextComponent,
  snapshotTestTextComponentAlign,
  snapshotTestTextComponentFontRamp,
  snapshotTestTextComponentFontWeight,
  snapshotTestTextComponentTransform,
  snapshotTestTextComponentTruncate,
  snapshotTestTextComponentVariant,
} from '../../../test/text'

test('A default Text component', () => {
  snapshotTestTextComponent(Text)
})

test('A Text component resized', () => {
  snapshotTestTextComponentFontRamp(Text)
})

test('A Text component weight', () => {
  snapshotTestTextComponentFontWeight(Text)
})

test('A Text component tuncated', () => {
  snapshotTestTextComponentTruncate(Text)
})

test('A Text component with variant', () => {
  snapshotTestTextComponentVariant(Text)
})

test('A Text component transformed', () => {
  snapshotTestTextComponentTransform(Text)
})

test('A Text component aligned', () => {
  snapshotTestTextComponentAlign(Text)
})
