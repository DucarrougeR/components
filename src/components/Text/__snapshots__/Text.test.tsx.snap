// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Text component resized 1`] = `
<p
  className=""
>
  ðŸ¥‘
</p>
`;

exports[`A Text component transformed 1`] = `
<p
  className=""
>
  ðŸ¥‘
</p>
`;

exports[`A Text component tuncated 1`] = `
<p
  className="undefined"
>
  ðŸ¥‘
</p>
`;

exports[`A Text component weight 1`] = `
<p
  className=""
>
  ðŸ¥‘
</p>
`;

exports[`A Text component with mode 1`] = `
<p
  className=""
>
  ðŸ¥‘
</p>
`;

exports[`A code Text component 1`] = `
<code
  className=""
>
  ðŸ¥‘
</code>
`;

exports[`A default Text component 1`] = `
<p
  className=""
>
  ðŸ¥‘
</p>
`;

exports[`A span Text component 1`] = `
<span
  className=""
>
  ðŸ¥‘
</span>
`;
